### PR TITLE
[WIP] feat: add tensorstore-zarr option  + create writers within MDAWidget

### PR DESCRIPTION
**Pull request by fdrgsp**
This PR:
- adds the `tensorstore-zarr` option to the `SaveGroupBox` in the `MDAWidget`.
- creates the writer directly in the `MDAWidget` (so that we will pass to the core `run_mda` method the full writer instead of just a path)

NOTE: this PR requires the `TensorStoreHandler` of `pymmcore-plus` which is not yet present in the current version.

related to https://github.com/pymmcore-plus/pymmcore-plus/pull/355.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -125,6 +125,9 @@ ignore = [
 [tool.ruff.format]
 docstring-code-format = true
 
+[tool.hatch.metadata]
+allow-direct-references = true
+
```
Comment (tlambert03) [inline pyproject.toml:130]:
```suggestion
```

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -27,5 +27,5 @@ repos:
       - id: mypy
         files: "^src/"
         additional_dependencies:
-          - pymmcore-plus >=0.9.0
+          - 'pymmcore-plus >=0.10.0'
```
Comment (tlambert03) [inline .pre-commit-config.yaml:30]:
```suggestion
          - pymmcore-plus >=0.10.2
```

File: src/pymmcore_widgets/mda/_core_mda.py [hunk 3]
```diff
@@ -201,16 +222,24 @@ def run_mda(self) -> None:
...
 
+        writer = None
         # technically, this is in the metadata as well, but isChecked is more direct
         if self.save_info.isChecked():
             save_path = self._update_save_path_from_metadata(
                 sequence, update_metadata=True
             )
-        else:
-            save_path = None
+            if isinstance(save_path, Path):
```
Comment (tlambert03) [inline src/pymmcore_widgets/mda/_core_mda.py:231]:
```suggestion
            if save_path is not None:
```

File: src/pymmcore_widgets/mda/_save_widget.py [hunk 4]
```diff
@@ -28,12 +28,14 @@ class SaveInfo(TypedDict):
         should_save: bool
 
 
+ZARR_TESNSORSTORE = "tensorstore-zarr"
```
Comment (tlambert03) [inline src/pymmcore_widgets/mda/_save_widget.py:31]:
```suggestion
ZARR_TENSORSTORE
```

File: tests/test_save_widget.py [hunk 5]
```diff
@@ -11,6 +11,7 @@
     SUBFOLDER,
     TIFF_SEQ,
     WRITERS,
+    ZARR_TESNSORSTORE,
```
Comment (tlambert03) [inline tests/test_save_widget.py:14]:
```suggestion
    ZARR_TENSORSTORE,
```

File: tests/test_useq_core_widgets.py [hunk 6]
```diff
@@ -6,16 +6,28 @@
...
 from pymmcore_widgets.mda._core_grid import CoreConnectedGridPlanWidget
-from pymmcore_widgets.mda._core_mda import CoreMDATabs
+from pymmcore_widgets.mda._core_mda import MAKE_WRITERS, CoreMDATabs
 from pymmcore_widgets.mda._core_positions import CoreConnectedPositionTable
 from pymmcore_widgets.mda._core_z import CoreConnectedZPlanWidget
+from pymmcore_widgets.mda._save_widget import (
+    OME_TIFF,
+    OME_ZARR,
+    TIFF_SEQ,
+    ZARR_TESNSORSTORE,
```
Comment (tlambert03) [inline tests/test_useq_core_widgets.py:29]:
```suggestion
    ZARR_TENSORSTORE,
```

File: src/pymmcore_widgets/mda/_core_mda.py [hunk 7]
```diff
@@ -26,7 +32,22 @@
...
+    OME_ZARR,
+    TIFF_SEQ,
+    ZARR_TESNSORSTORE,
+    SaveGroupBox,
+)
+
+MAKE_WRITERS: dict[str, Callable] = {
+    OME_TIFF: OMETiffWriter,
+    OME_ZARR: OMEZarrWriter,
+    ZARR_TESNSORSTORE: lambda path: TensorStoreHandler(
```
Comment (tlambert03) [inline src/pymmcore_widgets/mda/_core_mda.py:46]:
```suggestion
    ZARR_TENSORSTORE,
     SaveGroupBox,
 )

 MAKE_WRITERS: dict[str, Callable] = {
     OME_TIFF: OMETiffWriter,
     OME_ZARR: OMEZarrWriter,
     ZARR_TENSORSTORE: lambda path: TensorStoreHandler(
```

File: src/pymmcore_widgets/mda/_save_widget.py [hunk 8]
```diff
@@ -28,12 +28,14 @@ class SaveInfo(TypedDict):
...
 
 
+ZARR_TENSORSTORE = "tensorstore-zarr"
 OME_ZARR = "ome-zarr"
 OME_TIFF = "ome-tiff"
 TIFF_SEQ = "tiff-sequence"
 
 # dict with writer name and extension
 WRITERS: dict[str, list[str]] = {
+    ZARR_TESNSORSTORE: [".tensorstore.zarr"],
```
Comment (tlambert03) [inline src/pymmcore_widgets/mda/_save_widget.py:38]:
```suggestion
    ZARR_TENSORSTORE: [".tensorstore.zarr"],
```

File: tests/test_save_widget.py [hunk 9]
```diff
@@ -55,6 +55,24 @@ def test_set_get_value(qtbot: QtBot) -> None:
         "format": OME_ZARR,
     }
 
+    # setting zarr tensorstore format (dict)
+    wdg.setValue({"save_dir": str(path.parent), "save_name": "ts.tensorstore.zarr"})
+    assert wdg.value() == {
+        "save_dir": str(path.parent),
+        "save_name": "ts.tensorstore.zarr",
+        "should_save": False,
+        "format": ZARR_TESNSORSTORE,
```
Comment (tlambert03) [inline tests/test_save_widget.py:64]:
```suggestion
        "format": ZARR_TENSORSTORE,
```

File: tests/test_save_widget.py [hunk 10]
```diff
@@ -55,6 +55,24 @@ def test_set_get_value(qtbot: QtBot) -> None:
...
+    }
+
+    # setting zarr tensorstore format (path / string)
+    wdg.setValue("/some_path/ts.tensorstore.zarr")
+    assert wdg.value() == {
+        "save_dir": str(path.parent),
+        "save_name": "ts.tensorstore.zarr",
+        "should_save": True,
+        "format": ZARR_TESNSORSTORE,
```
Comment (tlambert03) [inline tests/test_save_widget.py:73]:
```suggestion
        "format": ZARR_TENSORSTORE,
```

File: tests/test_useq_core_widgets.py [hunk 11]
```diff
@@ -686,3 +697,21 @@ def test_get_next_available_paths_special_cases(tmp_path: Path) -> None:
     high = tmp_path / "test_12345.txt"
     high.touch()
     assert get_next_available_path(high).name == "test_12346.txt"
+
+
+data = [
+    ("./test.ome.tiff", OME_TIFF, OMETiffWriter),
+    ("./test.ome.zarr", OME_ZARR, OMEZarrWriter),
+    ("./test.tensorstore.zarr", ZARR_TESNSORSTORE, TensorStoreHandler),
```
Comment (tlambert03) [inline tests/test_useq_core_widgets.py:705]:
```suggestion
    ("./test.tensorstore.zarr", ZARR_TENSORSTORE, TensorStoreHandler),
```

**Review by tlambert03 — APPROVED** [review]
if this is working and doing what you need... i'm guess ok with it, but I'd like the `tensorstore.zarr` to be a short-lived extension.  basically, I don't want any users to say "what is a .tensorstore.zarr" file?  That's not actually a thing, and I definitely don't want to start a new file format convention... The fact that we're exposing a new-looking file object to end users is definitely problematic, and I'd like it if this was followed up on very soon

**Comment by fdrgsp** [comment]
> if this is working and doing what you need... i'm guess ok with it, but I'd like the `tensorstore.zarr` to be a short-lived extension. basically, I don't want any users to say "what is a .tensorstore.zarr" file? That's not actually a thing, and I definitely don't want to start a new file format convention... The fact that we're exposing a new-looking file object to end users is definitely problematic, and I'd like it if this was followed up on very soon

You mean that it should be simply zarr?

**Comment by tlambert03** [comment]
actually, can you make it very clear that this is an experimental feature (anywhere that an end-user would see this).  Perhaps in the widget itself, call it "Tensorstore (experimental)" or something

**Comment by tlambert03** [comment]
> You mean that it should be simply zarr?

no more than that.  If *we* are writing out the data for someone, they would very reasonably ask "what is the structure of this data"?  Right?  and this tensorstore writer is absolutely not writing a standardized specification, so now we are on the hook to explain/document/maintain/support this format.  that's baaaad... and definitely not something we want to get in the business of.  The answer to the question of "how is this data stored" should ideally always have a very clear answer: and ideally point to some external reference (like the ome-zarr specification or something like that).  This one has no specification and is completely mysterious.  For that reason, I'd almost prefer that this wasn't exposed at all (yet) ... but i also recognize that you're trying to use this in your own app, right?

**Comment by fdrgsp** [comment]
> but i also recognize that you're trying to use this in your own app, right?

yes I'm using it [here](https://github.com/fdrgsp/micromanager-gui) but I'm already using a modified version of `pymmcore-widgets` so if it is safer to wait until we define a better schema I'm ok with it! 👍🏻

**Comment by tlambert03** [comment]
ok great.  if you're not blocked, then let's hold off on exposing this just yet

**Pull request closed without merging**